packages/react-test-renderer/src/ReactShallowRenderer.js

```diff
@@ -561,19 +580,34 @@ class ReactShallowRenderer {
 
         this._mountClassComponent(element, this._context);
       } else {
-        const prevDispatcher = ReactCurrentDispatcher.current;
-        ReactCurrentDispatcher.current = this._dispatcher;
-        this._prepareToUseHooks(element.type);
-        try {
-          this._rendered = element.type.call(
-            undefined,
-            element.props,
-            this._context,
-          );
-        } finally {
-          ReactCurrentDispatcher.current = prevDispatcher;
+        let shouldRender = true;
+        if (
+          isMemo(element.type) &&
+          elementType === this._previousComponentIdentity &&
+          previousElement !== null
+        ) {
+          // This is a Memo component that is being re-rendered.
+          const compare = element.type.compare || shallowEqual;
+          if (compare(previousElement.props, element.props)) {
+            shouldRender = false;
+          }
+        }
+        if (shouldRender) {
+          const prevDispatcher = ReactCurrentDispatcher.current;
+          ReactCurrentDispatcher.current = this._dispatcher;
+          this._prepareToUseHooks(elementType);
+          try {
+            // elementType could still be a ForwardRef if it was
+            // nested inside Memo.
+            this._rendered =
+              elementType.$$typeof === ForwardRef
```

Comment:
should this use `isForwardRef(element)` (and cache the isForwardRef status in a var, since it seems to be checked multiple times)?

Reply:
It's a little confusing, but that doesn't work because `isForwardRef` expects a React Element for the `forwardRef`, not the actual object returned from `forwardRef` itself.

```js
const forwardRef = React.forwardRef(...);
isForwardRef(forwardRef) // false
isForwardRef(React.createElement(forwardRef)) // true
```